File: 2022/day16/B.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using vi = vector<int>;
using vll = vector<ll>;
template<typename t, size_t N>
using ar = array<t,N>;
using ii = ar<ll,2>;
using vii = vector<ii>;
using ld = long double;

#define all(v) begin(v), end(v)

const int INF = numeric_limits<int>::max();
const ll INFLL = numeric_limits<ll>::max();

mt19937_64 rng(chrono::steady_clock::now().time_since_epoch().count());
ll myRand(ll B) {
	return (unsigned ll)rng() % B;
}
template<typename t>
istream& operator >> (istream& in, vector<t>& vec) {
    for (t& x : vec)in >> x;
    return in;
}

template<typename t>
ostream& operator << (ostream& out, vector<t>& vec) {
    int n = (int)vec.size();
    for (int i = 0; i < n; ++i) {
        out << vec[i];
        if (i < n - 1)out << " ";
    }
    return out;
}

// t should support min-function (operator <)
template<typename t>
t min(const vector<t>& vec) {
    t ans = vec[0];
    for (const auto& el : vec) {
        ans = min(ans, el);
    }
    return ans;
}

// t should support max-function (operator <)
template<typename t>
t max(const vector<t>& vec) {
    t ans = vec[0];
    for (const auto& el : vec) {
        ans = max(ans, el);
    }
    return ans;
}
const int mxN = 3e4+3;
const ll K = 26;
using mp_t = unordered_map<ll,ll>;

ll n, m;
vll flow;
vector<vll> dist;
vll has_pos;

vector<vector<vector<vector<mp_t>>>> dp;
ll calc(ll u1, ll u2, ll t1, ll t2, ll mask) {
    if (min(t1,t2) >= K)return 0;
    t1 = min(t1,K);
    t2 = min(t2,K);
    if (u1 > u2) {
        swap(u1,u2);
        swap(t1,t2);
    }

    if (dp[u1][u2][t1][t2].count(mask))return dp[u1][u2][t1][t2][mask];

    ll ret = 0;
    for (ll i = 0; i < m; ++i) {
        if (mask & (1LL<<i))continue;
        for (ll j = 0; j < m; ++j) if (j != i) {
            if (mask & (1LL<<j))continue;
            ll there1 = t1 + dist[u1][has_pos[i]] + 1;
            ll there2 = t2 + dist[u2][has_pos[j]] + 1;

            ll c = flow[has_pos[i]] * max(0LL, K - there1) + flow[has_pos[j]] * max(0LL, K - there2);
            c += calc(has_pos[i], has_pos[j], there1, there2, mask | (1LL<<i) | (1LL << j));
            ret = max(ret,c);
        }
    }
    dp[u1][u2][t1][t2][mask] = ret;
    return ret;
}

void testCase() {
    cin >> n;
    ll start;
    cin >> start;

    flow = vll(n);
    cin >> flow;

    for (ll i = 0; i < n; ++i)if (flow[i])has_pos.push_back(i);
    m = has_pos.size();

    dist = vector<vll>(n,vll(n, 0));

    for (int i = 0; i < n * n; ++i) {
        ll u,v,d;
        cin >> u >> v >> d;
        dist[u][v] = d;
        dist[v][u] = d;
    }

    dp = vector<vector<vector<vector<mp_t>>>>(n,vector<vector<vector<mp_t>>>(n,vector<vector<mp_t>>(K+1,vector<mp_t>(K+1, mp_t()))));
    cout << calc(start, start, 0, 0, 0) << endl;
}

void setIO() {
    cin.tie(0)->sync_with_stdio(0);
}

void pre() {

}

int main() {
    setIO();
    pre();
    int t = 1;
    //cin >> t;

    for (int tc = 1; tc <= t; ++tc) {
        testCase();
    }
    return 0;
}

File: 2022/day16/B.py
from itertools import permutations
import sys
with open(sys.argv[1], "r") as f:
    lines = list(map(lambda l: l.split(), f.readlines()))

flow = dict()
adj = dict()

has_pos = []

for line in lines:
    key = line[1]
    flow[key] = int(line[4][5:-1])

    if flow[key] > 0:
        has_pos.append(key)

    if key not in adj:
        adj[key] = []

    for valc in line[9:]:
        nxt = valc
        if valc != line[-1]:
            nxt = valc[:-1]

        adj[key].append(nxt)


m = len(has_pos)
dist = dict()

for i in adj:
    dist[i] = dict()
    dist[i][i] = 0

for i in adj:
    for j in adj[i]:
        dist[i][j] = 1

for k in adj:
    for i in adj:
        for j in adj:
            if i == j:
                continue
            if k in dist[i] and j in dist[k]:
                if j not in dist[i]:
                    dist[i][j] = dist[i][k] + dist[k][j]
                dist[i][j] = min(dist[i][j], dist[i][k]+dist[k][j])


toint = dict()
ptr = 0
start = -1
for u in dist:
    toint[u] = ptr
    if (u == "AA"):
        start = ptr
    ptr += 1

print(len(dist))
print(start)

for i in dist:
    print(flow[i])

for u in dist:
    for v in dist:
        print(f"{toint[u]} {toint[v]} {dist[u][v]}")

        
dp = dict()

def calc(key1,key2, t1, t2, mask):
    if (key1,key2,t1,t2,mask) in dp:
        return dp[(key1,key2,t1,t2,mask)]
    if min(t1,t2) >= 26:
        return 0

    ret = 0

    for i in range(m):
        if mask & (1<<i):
            continue
        for j in range(m):
            if j == i:
                continue
            if mask & (1<<j):
                continue

            there1 = t1 + dist[key1][has_pos[i]] + 1
            there2 = t2 + dist[key2][has_pos[j]] + 1

            c = flow[has_pos[i]] * max(0, 26 - there1) + flow[has_pos[j]] * max(0,26 - there2)
            c += calc(has_pos[i], has_pos[j], there1,there2, mask | (1<<i) | (1<<j))
            ret = max(ret,c)

    dp[(key1,key2,t1,t2,mask)] = ret
    return ret

#print(calc("AA","AA",0,0,0))
